packages/plugin-ext/src/plugin/telemetry-ext.ts

```diff
@@ -0,0 +1,373 @@
+            updatedStack += stack.substring(lastIndex, result.index) + '<REDACTED: user-file-path>';
+            lastIndex = fileRegex.lastIndex;
+        }
+    }
+    if (lastIndex < stack.length) {
+        updatedStack += stack.substr(lastIndex);
+    }
+
+    return updatedStack;
+}
+
+/**
+ * Attempts to remove commonly leaked PII
+ * @param property The property which will be removed if it contains user data
+ * @returns The new value for the property
+ */
+function removePropertiesWithPossibleUserInfo(property: string): string {
+    // If for some reason it is undefined we skip it (this shouldn't be possible);
+    if (!property) {
+        return property;
+    }
+
+    const value = property.toLowerCase();
+
+    const userDataRegexes = [
+        { label: 'Google API Key', regex: /AIza[0-9A-Za-z-_]{35}/ },
+        { label: 'Slack Token', regex: /xox[pbar]\-[A-Za-z0-9]/ },
+        { label: 'Generic Secret', regex: /(key|token|sig|secret|signature|password|passwd|pwd|android:value)[^a-zA-Z0-9]/ },
+        { label: 'Email', regex: /@[a-zA-Z0-9-]+\.[a-zA-Z0-9-]+/ } // Regex which matches @*.site
+    ];
+
+    // Check for common user data in the telemetry events
+    for (const secretRegex of userDataRegexes) {
+        if (secretRegex.regex.test(value)) {
+            return `<REDACTED: ${secretRegex.label}>`;
+        }
+    }
+
+    return property;
+}
+
+/**
+ * Does a best possible effort to clean a data object from any possible PII.
+ * @param data The data object to clean
+ * @param paths Any additional patterns that should be removed from the data set
+ * @returns A new object with the PII removed
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function cleanData(data: Record<string, any>, cleanUpPatterns: RegExp[]): Record<string, any> {
+    return cloneAndChange(data, value => {
+
+        // If it's a trusted value it means it's okay to skip cleaning so we don't clean it
+        if (value instanceof TelemetryTrustedValue) {
+            return value.value;
+        }
+
+        // We only know how to clean strings
+        if (typeof value === 'string') {
+            let updatedProperty = value.replace(/%20/g, ' ');
+
+            // First we anonymize any possible file paths
+            updatedProperty = anonymizeFilePaths(updatedProperty, cleanUpPatterns);
+
+            // Then we do a simple regex replace with the defined patterns
+            for (const regexp of cleanUpPatterns) {
+                updatedProperty = updatedProperty.replace(regexp, '');
+            }
+
+            // Lastly, remove commonly leaked PII
+            updatedProperty = removePropertiesWithPossibleUserInfo(updatedProperty);
+
+            return updatedProperty;
+        }
+        return undefined;
+    }, new Set());
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function cloneAndChange(obj: any, changer: (orig: any) => any, seen: Set<any>): any {
```

Comment:
This is likely better suited to packages/core/src/common/objects.ts?

Reply:
This was moved in the core/common/object file, it is definitely a better place! I also transfered the "clone" comment.